File: Sorting_Techniques/quick_sort/QuickSort.cpp
// C++ program to implement Quick Sort

#include <iostream>
using namespace std;

       //function to swap two elements

		void swap(int *a, int *b)
		{
			int t = *a;
			*a = *b;
			*b = t;
		}
		
		/*A  function  that takes last element as pivot, places 
           the pivot element at its correct position in sorted 
             array, and places all smaller (smaller than pivot) 
              to left of pivot and all greater elements to right of pivot */
		
		int partition(int ar[], int p, int r)
		{
			int x,i,j;
			
			x = ar[r]; //pivot element
			i = p-1;
			
			for(j = p; j <= r-1; j++)
			{
				if(ar[j] <= x)
				{
					i++;
					
					 swap(&ar[i], &ar[j]);
				}
			}
			
			swap(&ar[i+1], &ar[r]);
			
			return (i+1);
		}
		
		//function to implement QuickSort
		
		void QuickSort(int ar[], int p, int r)
		{
			if(p<r)
			{
				int q = partition(ar, p, r);
				
				//sorting elements before and after partition
				
				QuickSort(ar, p, q-1);
				QuickSort(ar, q+1, r);
			}
		}
		
		//function to display elements of the array
		
		void displayArray(int ar[], int len)
		{
			for(int i = 0; i < len; i++)
			 
			   cout<<ar[i]<<"\t";
   
		}
		
		 //main function to control program execution by calling other functions
		
		int main()
		{
			int arr[10], size;
			
			cout<<"Enter the length of the array"<<endl;
			 
			cin>>size;
			
			cout<<"Enter the elements in an array "<<endl;
			
			 for(int i = 0; i < size; i++ )
			     cin>>arr[i];
			     
			  QuickSort(arr, 0, size-1); 
			  
            cout<<"Sorted array: n"<<endl; 
             
             displayArray(arr, size); 
             
       return 0;    
		}

File: Sorting_Techniques/quick_sort/quicksort.js
function swap(items, leftIndex, rightIndex){
    var temp = items[leftIndex];
    items[leftIndex] = items[rightIndex];
    items[rightIndex] = temp;
}

function partition(items, left, right) {
    var pivot = items[Math.floor((right + left) / 2)];
    var i = left;
    var j = right;

    while (i <= j) {
        while (items[i] < pivot) {
            i++;
        }
        while (items[j] > pivot) {
            j--;
        }
        if (i <= j) {
            swap(items, i, j); 
            i++;
            j--;
        }
    }
    return i;
}

function quickSort(items, left, right) {
    var index;
    if (items.length > 1) {
        index = partition(items, left, right); 
        if (left < index - 1) { 
            quickSort(items, left, index - 1);
        }
        if (index < right) { 
            quickSort(items, index, right);
        }
    }
    return items;
}

var arr = [12, 321, 45, 8, 2, 0];
var sortedArray = quickSort(arr, 0, arr.length - 1);
console.log(sortedArray);
